src/components/checkbox/GroupableCheckboxReducers.ts

```diff
@@ -0,0 +1,171 @@
+                                         action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      if (action.payload.isParent) {
+        return addParentCheckbox(state, action);
+      }
+      return addChildCheckbox(state, action);
+    case GroupableCheckboxActions.removeGroup:
+      if (state.parentId === action.payload.parentId) {
+        let isChecked: boolean = false;
+        const newCheckboxes = _.reject(state.checkboxes, (checkbox: ICheckboxState) => {
+          const reject: boolean = checkbox.id === action.payload.id;
+          isChecked = reject ? checkbox.checked : isChecked;
+          return reject;
+      });
+
+        const total: number = state.total - 1;
+        const nbSelected: number = isChecked ? state.nbSelected - 1 : state.nbSelected;
+        return {
+          ...state,
+          checkboxes: newCheckboxes,
+          nbSelected,
+          total,
+        };
+      }
+    case GroupableCheckboxActions.toggleGroup:
+      if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+        let isChecked: boolean;
+        if (action.payload.isParent) {
+          isChecked = !state.parent.checked;
+          return {
+            ...state,
+            parent: {
+              id: action.payload.id,
+              checked: isChecked,
+            },
+            checkboxes: _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+              checkbox.checked = isChecked;
+              return checkbox;
+            }),
+            nbSelected: isChecked ? state.total : 0,
+          };
+        }
+        const checkboxes: ICheckboxState[] = _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+          if (checkbox.id === action.payload.id) {
+            isChecked = !checkbox.checked;
+            return {
+              id: checkbox.id,
+              checked: isChecked,
+            };
+          }
+          return checkbox;
+        });
+        if (checkboxes) {
+          const nbSelected: number = state.nbSelected += (isChecked ? 1 : -1);
+          return {
+            ...state,
+            parent: {
+              id: action.payload.parentId,
+              checked: nbSelected === state.total,
+            },
+            nbSelected,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
+      return state;
+    default:
+      return state;
+  }
+};
+
+export const groupableCheckboxesReducer = (state: IGroupableCheckboxesState[] = groupableCheckboxesInitialState,
+                                           action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState[] => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      const groupableCheckboxesState: IGroupableCheckboxesState = _.findWhere(state, (groupCheckboxState: IGroupableCheckboxesState) =>
```

Comment:
Why don't you have another reducer for the groupable checkbox? You have one for the collection, you should have one for the object. So that your reducers don't because too complicated.

Reply:
Nevermind, you already have one (`groupableCheckboxReducer`). Just use it instead.